### PR TITLE
Split Retrievable Resolver into more actors

### PR DESCRIPTION
**Pull request by jmsfltchr**
## What is the goal of this PR?

We start to devolve bloated actors in the reasoner to offload work to more actors. We start with actors that resolve retrievables, splitting the work for each set of bounds into a new actor, with one actor to coordinate them.

## What are the changes implemented in this PR?

- We split the content of RetrievableResolver so that what were RequestStates are now separate actors.
- We keep subsumption active by creating specialised request types to indicate that subsumption should be used. The RetrievableResolver actor coordinates subsumption, and directs the appropriate requests to the BoundRetrievableResolver actors.

Part of #6356

### CODE REVIEW

File: common/poller/FilteredPoller.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+
+package com.vaticle.typedb.core.common.poller;
+
+import java.util.Optional;
+import java.util.function.Predicate;
+
+public class FilteredPoller<T> extends AbstractPoller<T> {
```
Comment (jmsfltchr) [inline common/poller/FilteredPoller.java:24]: In fact we don't need the `FilteredPoller` in the end - think we should keep it anyway?

File: reasoner/resolution/answer/AnswerStateImpl.java [hunk 2]
```diff
@@ -1003,6 +1003,11 @@ public P aggregateToUpstream(ConceptMap concepts) {
                 return parent().with(concepts);
             }
 
+            @Override
+            public Retrievable<P> with(ConceptMap extension) {
+                return new RetrievableImpl<>(filter, parent(), extendAnswer(extension), root());
+            }
+
```
Comment (jmsfltchr) [inline reasoner/resolution/answer/AnswerStateImpl.java:1010]: Needed so that we can pass from a subsumer to the resolver it subsumes without the answerstate following the normal `upstream` pattern.

File: reasoner/resolution/framework/AnswerCache.java [hunk 3]
```diff
@@ -37,42 +37,51 @@
...
     protected boolean complete;
     protected final Map<SUBSUMES, ? extends AnswerCache<?, SUBSUMES>> cacheRegister;
     protected final ConceptMap state;
+    private final Supplier<FunctionalIterator<ANSWER>> answerSourceSupplier;
+    private boolean sourceCleared;
+    private boolean sourceExhausted;
 
-    protected AnswerCache(Map<SUBSUMES, ? extends AnswerCache<?, SUBSUMES>> cacheRegister, ConceptMap state) {
+    protected AnswerCache(Map<SUBSUMES, ? extends AnswerCache<?, SUBSUMES>> cacheRegister, ConceptMap state,
+                          Supplier<FunctionalIterator<ANSWER>> answerSourceSupplier) {
```
Comment (jmsfltchr) [inline reasoner/resolution/framework/AnswerCache.java:62]: We only ever use one source, and it's always an iterator, so that lets us simplify. Mainly doing this because we need it. It means that we are able to kill the traversal when we do subsumption, where previously we got the cache to set itself as complete because it was aware of the other caches.

File: reasoner/resolution/framework/Request.java [hunk 4]
```diff
@@ -95,4 +98,86 @@ public String toString() {
...
+
+    public boolean isToSubsumer() {
+        return false;
+    }
+
+    public ToSubsumer asToSubsumer() {
+        throw TypeDBException.of(ILLEGAL_STATE);
+    }
+
+    public static class ToSubsumed extends Request {
```
Comment (jmsfltchr) [inline reasoner/resolution/framework/Request.java:117]: These are the extra request types that can tell a resolver where to go next for subsumption answers

File: reasoner/resolution/ResolverRegistry.java [hunk 5]
```diff
@@ -188,6 +191,16 @@ public ResolverView registerResolvable(Resolvable<?> resolvable) {
         return ResolverView.retrievable(resolver, retrievable.retrieves());
     }
 
+    public Actor.Driver<BoundRetrievableResolver> registerBoundRetrievable(Retrievable retrievable, ConceptMap bounds) {
```
Comment (jmsfltchr) [inline reasoner/resolution/ResolverRegistry.java:194]: We need an elegant way to terminate these

File: reasoner/resolution/resolver/ConcludableResolver.java [hunk 6]
```diff
@@ -255,7 +258,7 @@ private void nextAnswer(Request fromUpstream, CachingRequestState<?, ConceptMap>
                     ConcludableAnswerCache answerCache = new ConcludableAnswerCache(cacheRegister, answerFromUpstream);
                     cacheRegister.put(answerFromUpstream, answerCache);
                     requestState = new Explain(fromUpstream, answerCache, iteration);
-                    registerRules(fromUpstream, requestState.asExploration());
+                    requestState.asExploration().downstreamManager().addDownstreams(ruleDownstreams(fromUpstream));
```
Comment (jmsfltchr) [inline reasoner/resolution/resolver/ConcludableResolver.java:261]: Just some tidying in this class

File: reasoner/resolution/resolver/RetrievableResolver.java [hunk 7]
```diff
@@ -69,89 +64,74 @@ public void receiveRequest(Request fromUpstream, int iteration) {
...
+            Driver<BoundRetrievableResolver> boundRetrievable = getOrReplaceBoundRetrievable(root, bounds);
+            // TODO: Re-enable subsumption when async bug is fixed
+            // Optional<ConceptMap> subsumer = subsumptionTrackers.computeIfAbsent(
+            //         root, r -> new SubsumptionTracker()).getSubsumer(bounds);
+            // // If there is a finished subsumer, let the BoundRetrievable know that it can go there for answers
+            // Request request = subsumer
+            //         .map(conceptMap -> Request.ToSubsumed.create(
+            //                 driver(), boundRetrievable, boundRetrievablesByRoot.get(root).get(conceptMap),
+            //                 fromUpstream.partialAnswer()))
+            //         .orElseGet(() -> Request.create(driver(), boundRetrievable, fromUpstream.partialAnswer()));
```
Comment (jmsfltchr) [inline reasoner/resolution/resolver/RetrievableResolver.java:84]: How subsumption works when enabled

File: reasoner/resolution/resolver/SubsumptionTracker.java [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+import com.vaticle.typedb.core.concept.answer.ConceptMap;
+import com.vaticle.typedb.core.traversal.common.Identifier;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+class SubsumptionTracker {
```
Comment (jmsfltchr) [inline reasoner/resolution/resolver/SubsumptionTracker.java:30]: If we do manage to create an abstract `Coordinator` class, perhaps with a subclass which is a `SubsumptiveCoordinator` then `SubsumptionTracker` can be an inner class there. For now it's a bit random where it sits, happy enough for that while we change things.

File: reasoner/resolution/resolver/BoundRetrievableResolver.java [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+    @Override
+    public void receiveRequest(Request fromUpstream, int iteration) {
+        if (fromUpstream.isToSubsumed()) {
+            assert fromUpstream.partialAnswer().conceptMap().equals(bounds);
+            receiveSubsumedRequest(fromUpstream.asToSubsumed(), iteration);
+        } else if (fromUpstream.isToSubsumer()) {
+            receiveSubsumerRequest(fromUpstream.asToSubsumer(), iteration);
+        } else {
+            assert fromUpstream.partialAnswer().conceptMap().equals(bounds);
+            receiveDirectRequest(fromUpstream, iteration);
```
Comment (jmsfltchr) [inline reasoner/resolution/resolver/BoundRetrievableResolver.java:70]: This is the (not impressively named) path for non-subsumptive requests

File: reasoner/resolution/resolver/BoundRetrievableResolver.java [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+                fromUpstream, request -> new SubsumerRequestState(request, cache, iteration)));
+    }
+
+    private void receiveDirectRequest(Request fromUpstream, int iteration) {
+        sendAnswerOrFail(fromUpstream, iteration, requestStates.computeIfAbsent(
+                fromUpstream, request -> new BoundRequestState(request, cache, iteration)));
+    }
+
+    @Override
+    protected void receiveAnswer(Response.Answer fromDownstream, int iteration) {
```
Comment (jmsfltchr) [inline reasoner/resolution/resolver/BoundRetrievableResolver.java:102]: Only for subsumption

File: reasoner/resolution/resolver/BoundRetrievableResolver.java [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+            if (upstreamAnswer.isPresent()) {
+                answerToUpstream(upstreamAnswer.get(), fromUpstream, iteration);
+            } else {
+                requestFromSubsumer(fromUpstream, iteration);
+            }
+        }
+    }
+
+    @Override
+    protected void receiveFail(Response.Fail fromDownstream, int iteration) {
```
Comment (jmsfltchr) [inline reasoner/resolution/resolver/BoundRetrievableResolver.java:117]: Only for subsumption

File: reasoner/resolution/framework/AnswerCache.java [hunk 12]
```diff
@@ -37,42 +37,51 @@
...
+    protected FunctionalIterator<ANSWER> answerSource;
     protected boolean complete;
     protected final Map<SUBSUMES, ? extends AnswerCache<?, SUBSUMES>> cacheRegister;
     protected final ConceptMap state;
+    private final Supplier<FunctionalIterator<ANSWER>> answerSourceSupplier;
+    private boolean sourceCleared;
+    private boolean sourceExhausted;
 
-    protected AnswerCache(Map<SUBSUMES, ? extends AnswerCache<?, SUBSUMES>> cacheRegister, ConceptMap state) {
+    protected AnswerCache(Map<SUBSUMES, ? extends AnswerCache<?, SUBSUMES>> cacheRegister, ConceptMap state,
```
[the remark this answers is not in the document]
Comment (flyingsilverfin) [reply reasoner/resolution/framework/AnswerCache.java:61]: we decided that we will pass over this for now and see how it evolves over the refactor

File: reasoner/resolution/resolver/RetrievableResolver.java [hunk 13]
```diff
@@ -69,89 +64,74 @@ public void receiveRequest(Request fromUpstream, int iteration) {
...
 
-    @Override
-    protected void receiveAnswer(Answer fromDownstream, int iteration) {
-        throw TypeDBException.of(ILLEGAL_STATE);
-    }
-
-    @Override
-    protected void receiveFail(Response.Fail fromDownstream, int iteration) {
-        throw TypeDBException.of(ILLEGAL_STATE);
+    private Driver<BoundRetrievableResolver> getOrReplaceBoundRetrievable(Driver<? extends Resolver<?>> root, ConceptMap bounds) {
```
Comment (flyingsilverfin) [inline reasoner/resolution/resolver/RetrievableResolver.java:98]: replace?

File: reasoner/resolution/resolver/SubsumptionTracker.java [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+        this.finishedStates = new HashSet<>();
+        this.subsumersMap = new HashMap<>();
+        this.finishedMapping = new HashMap<>();
+    }
+
+    public void addFinished(ConceptMap conceptMap) {
+        this.finishedStates.add(conceptMap);
+    }
+
+    public Optional<ConceptMap> getSubsumer(ConceptMap unfinished) {
```
Comment (flyingsilverfin) [inline reasoner/resolution/resolver/SubsumptionTracker.java:46]: `getSubsumer` doesn't sound like it has any notion of finish/unfinished. 
This class from outside appears to have only to do with concept maps being subsets of other ones, right?

Maybe you already want to turn this into a `SubsumptiveCoordinator` which possibly has a `SubsumptionTracker` in it? The `Tracker` only knows about relationships between concept maps. 
The coordinator keeps track of finished/unfinished states and is roughly aware of "resolvers" answering queries and being complete or incomplete.

In other words the story/abstraction still feels a bit rough. How would you explain it in words?

**Review by flyingsilverfin — CHANGES_REQUESTED** [review]
I'm a bit conflicted about how to proceed - i can't quite keep up with your reviews, especially since I'm remote right now and it probably warrants discussion. I'd probably say it's better to discuss the model you're implementing and then just do it, and we can do a pair review at certain checkpoints?

File: reasoner/resolution/ResolverRegistry.java (hunk 5 above)
Comment (jmsfltchr) [reply reasoner/resolution/ResolverRegistry.java:194]: Ah yes, currently we don't terminate the bound resolvers when we're actually done with them, because they are bound to a root and iteration.

File: reasoner/resolution/resolver/RetrievableResolver.java (hunk 13 above)
Comment (jmsfltchr) [reply reasoner/resolution/resolver/RetrievableResolver.java:98]: Yes I think so...

File: reasoner/resolution/resolver/RetrievableResolver.java [hunk 15]
```diff
@@ -69,89 +64,74 @@ public void receiveRequest(Request fromUpstream, int iteration) {
...
+            requestMapByRoot.computeIfAbsent(root, r -> new HashMap<>()).put(request, fromUpstream);
+            requestFromDownstream(request, fromUpstream, iteration);
         }
     }
 
     private void prepareNextIteration(Driver<? extends Resolver<?>> root, int iteration) {
         iterationByRoot.put(root, iteration);
-        cacheRegistersByRoot.get(root).clear();
+        boundRetrievablesByRoot.remove(root);
+        subsumptionTrackers.remove(root);
```
[the remark this answers is not in the document]
Comment (jmsfltchr) [reply reasoner/resolution/resolver/RetrievableResolver.java:94]: So we've decided we're keeping this for as long as it isn't getting in the way, since it nearly works.

File: reasoner/resolution/resolver/SubsumptionTracker.java (hunk 14 above)
Comment (jmsfltchr) [reply reasoner/resolution/resolver/SubsumptionTracker.java:46]: The `SubsumptionTracker` class isn't doing much if it isn't tracking which subsumers are finished and which finished subsumers we've already found in the past. I've moved it to be an inner class and tweaked `getSubsumer` to `getFinishedSubsumer` for clarity. Unsure of the next move.
